src/gadgets/isnonzero.cpp

```diff
@@ -0,0 +1,60 @@
+#include "gadgets/isnonzero.hpp"
+#include "utils.hpp"
+#include <libsnark/gadgetlib1/gadgets/basic_gadgets.hpp>
+
+using libsnark::generate_boolean_r1cs_constraint;
+
+namespace ethsnarks {
+
+
+IsNonZero::IsNonZero(
+	ProtoboardT& in_pb,
+    const VariableT& in_var,
+    const std::string &annotation_prefix
+) :
+	GadgetT(in_pb, annotation_prefix),
+	m_X(in_var),
+	m_Y(make_variable(in_pb, FMT(this->annotation_prefix, ".Y"))),
+	m_M(make_variable(in_pb, FMT(this->annotation_prefix, ".M")))
+{
+
+}
+
+
+const VariableT& IsNonZero::result() const
+{
+	return this->m_Y;
+}
+
+
+void IsNonZero::generate_r1cs_constraints()
+{
+	generate_boolean_r1cs_constraint<FieldT>(pb, m_Y);
```

Comment:
The boolean constraint is needed for `Y` because `M` is specified as a witness and not validated on its own - if Y is forced to be 0 or 1, then the only valid value for `M` is `1/X`